src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/segmentation/KernelSegmenter.java

```diff
@@ -0,0 +1,526 @@
+            if (windowSize > data.size()) {
+                logger.warn(String.format("Number of points needed to calculate local changepoint costs (2 * window size = %d) " +
+                        "exceeds number of data points %d.  Local changepoint costs will not be calculated for this window size.",
+                        2 * windowSize, data.size()));
+                continue;
+            }
+            final double[] windowCosts = calculateWindowCosts(reducedObservationMatrix, kernelApproximationDiagonal, windowSize);
+
+            logger.debug(String.format("Finding local minima of local changepoint costs for window size %d...", windowSize));
+            final List<Integer> windowCostLocalMinima = new PersistenceOptimizer(windowCosts).getMinimaIndices();
+            windowCostLocalMinima.remove(Integer.valueOf(0));                //remove first data point if present
+            windowCostLocalMinima.remove(Integer.valueOf(data.size() - 1));  //remove last data point if present
+            changepointCandidates.addAll(windowCostLocalMinima.subList(0, Math.min(maxNumChangepoints, windowCostLocalMinima.size())));
+        }
+
+        if (changepointCandidates.isEmpty()) {
+            throw new UserException.BadInput("No changepoint candidates found.  Window sizes may be inappropriate.");
+        }
+
+        return changepointCandidates;
+    }
+
+    //performs backwards model selection to order changepoints by increasing change to the global segmentation cost
+    //and returns the requested number
+    private static List<Integer> selectChangepoints(final List<Integer> changepointCandidates,
+                                                    final int maxNumChangepoints,
+                                                    final double numChangepointsPenaltyLinearFactor,
+                                                    final double numChangepointsPenaltyLogLinearFactor,
+                                                    final RealMatrix reducedObservationMatrix,
+                                                    final double[] kernelApproximationDiagonal) {
+        final List<Integer> changepoints = new ArrayList<>(changepointCandidates.size());
+
+        //calculate penalties as a function of the number of changepoints
+        final int numData = reducedObservationMatrix.getRowDimension();
+        final List<Double> changepointPenalties = IntStream.range(0, maxNumChangepoints + 1).boxed()
+                .map(numChangepoints -> numChangepointsPenaltyLinearFactor * numChangepoints
+                        + numChangepointsPenaltyLogLinearFactor * numChangepoints * Math.log(numData / (numChangepoints + EPSILON)))
+                .collect(Collectors.toList());
+
+        //construct initial list of all segments and initialize costs
+        final List<Integer> candidateStarts = changepointCandidates.stream().sorted().distinct()
```

Comment:
How about combine starts and ends into a `List<IndexRange> candidate segments`, which could be created from a single stream?

Reply:
Segments as I've defined them are end inclusive, but `IndexRange` is supposed to represent an end exclusive interval, so I'll leave things as they are.